<a href="https://colab.research.google.com/github/MatiasRolon/BDM/blob/master/BDM_TP05_01.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [0]:

#Importamos las librerías que usaremos en todos los puntos.
#Ejecutar este codigo antes que cualquier punto.
from sklearn.datasets import load_wine
from sklearn import tree
from sklearn.model_selection import train_test_split
from sklearn import metrics
from sklearn import preprocessing
from sklearn.metrics import precision_recall_fscore_support
import graphviz
import pandas as pd 
import numpy as np



---

## PUNTO 1
**Consigna:** A partir del dataset presentado a continuación, y teniendo en cuenta las fórmulas de entropía y ganancia de información, calcule y diagrame el árbol de decisión que le permita decidir si comer asado o no en función del clima:

In [0]:
!wget https://raw.githubusercontent.com/jumafernandez/BDM/master/TPs/TP05/TP0501/db-asado.xls
# Leemos el archivo
data1 = pd.read_excel("db-asado.xls") 
# Preview the first 5 lines of the loaded data 
data1.head()

In [0]:
le = preprocessing.LabelEncoder()
for column_name in data1.columns:
  if (data1[column_name].dtype == object) & (column_name!='ASADO'):
    data1[column_name] = le.fit_transform(data1[column_name])
    
feature_names = list(data1.columns)
feature_names.remove('ASADO')
x1 = data1[feature_names]
y1 = data1.ASADO
le.fit(data1['ASADO'])
target_names=le.classes_

arbol1 = tree.DecisionTreeClassifier(criterion='entropy')
arbol1 = arbol1.fit(x1, y1)

dot_data = tree.export_graphviz(arbol1, out_file=None,
                                feature_names=feature_names,
                                class_names=target_names,
                                label='all',
                                filled=True, rounded=True,
                                special_characters=True)  

# Capturo el source
graph = graphviz.Source(dot_data)
# Muestro el contenido de graph
graph



---


## PUNTO 2
**Consigna: Trabaje con el dataset de Scikit Learn “wine”:**

In [0]:

wine= load_wine()
# lo visualizamos
print(wine.DESCR)

**2.a Utilice el metadata que provee la librería, ¿Cuál es el tema que
aborda el dataset?**

El dataset aborda las caracteristicas de distintos tipos de vinos, como por ejemplo su alcalinidad, acido malico, fenoles, su color, tono etc.

**2.b Genere el árbol de decisión que permita clasificar los diferentes tipos de vino utilizando un muestreo con proporciones de 80% para entrenamiento y 20% para testeo.**

Si no la tenemos instalada, descargamos la libreria para graficar el arbol.

In [0]:
!pip install graphviz

# Separo en 80-20 entrenamiento y testeo
X_train, X_test, y_train, y_test = train_test_split(wine.data, wine.target, random_state=0, test_size=0.2)
# Creo el objeto Decision Tree classifer
arbol_tt = tree.DecisionTreeClassifier(criterion = "entropy")
# Entreno el Decision Tree Classifer
arbol_tt = arbol_tt.fit(X_train,y_train)
#Realizo las predicciones en función del árbol generado
y_pred = arbol_tt.predict(X_test)
# graficamos el arbol
dot_data = tree.export_graphviz(arbol_tt, out_file=None,
                                feature_names=wine.feature_names,
                                class_names=wine.target_names,
                                filled=True, rounded=True,
                                special_characters=True)  

# Capturo el source
graph = graphviz.Source(dot_data)
# Muestro el contenido de graph
graph

**2.c Explore la solución dada y las posibles configuraciones para obtener
un nuevo árbol que clasifique “mejor”. Documente las conclusiones.**

In [9]:
# Vamos a testear el modelo recien hecho.
print("Accuracy:",metrics.accuracy_score(y_test, y_pred))

Accuracy: 0.9444444444444444


Acabamos de ver la precision para el modelo anteriormente planteado. Podemos observar que da un 91.6% de exactitud de tuplas correctamente clasificadas.

Para obtener un mejor arbol podemos cambiarle uno de los parametros principales a la hora de hacer el arbol: el criterio. En este caso lo modificaremos para que en vez de ser la Entropía, sea Gini

In [10]:
X_train, X_test, y_train, y_test = train_test_split(wine.data, wine.target, random_state=0, test_size=0.2)
#criterio modificado
arbol_tt = tree.DecisionTreeClassifier(criterion = "gini")
arbol_tt = arbol_tt.fit(X_train,y_train)
y_pred = arbol_tt.predict(X_test)
# Vamos a testear el modelo
print("Accuracy:",metrics.accuracy_score(y_test, y_pred))

Accuracy: 0.9722222222222222


Asi, hemos encontrado una configuración que refleja un mayor porcentaje de exactitud, con el 97.2% de tuplas correctamente clasificadas.



---


## PUNTO 3
**Consiga: Ahora, analice el archivo zoo.csv:**

In [0]:
!rm zoo.csv
!wget https://raw.githubusercontent.com/jumafernandez/BDM/master/TPs/TP05/TP0501/zoo.csv

# Lee el archivo
data3 = pd.read_csv("zoo.csv") 
# Preview the first 5 lines of the loaded data 
data3.head()

**3 a. Genere el árbol de decisión que permita inferir el tipo de animal en función de sus características. Explique someramente que resultado
se obtiene en términos del árbol y en términos de la eficiencia del
mismo.**

In [0]:
le = preprocessing.LabelEncoder()
for column_name in data3.columns:
  if (data3[column_name].dtype == object) & (column_name!='type'):
    data3[column_name] = le.fit_transform(data3[column_name])

# Features
feature_names = list(data3.columns)
# Elimino survived porque es la clase
feature_names.remove('type')
x3 = data3[feature_names]
# Target
y3 = data3.type
le.fit(data3['type'])
target_names=le.classes_
#armo el arbol separando en 80-20 entrenamiento y testeo
X_train, X_test, y_train, y_test = train_test_split(x3, y3, random_state=0, test_size=0.2)
# Creo el objeto Decision Tree classifer
arbol3 = tree.DecisionTreeClassifier(criterion = "entropy")
# Entreno el Decision Tree Classifer
arbol3 = arbol3.fit(X_train,y_train)
#Realizo las predicciones en función del árbol generado
y_pred = arbol3.predict(X_test)
# graficamos el arbol
dot_data = tree.export_graphviz(arbol3, out_file=None,
                                feature_names=feature_names,
                                class_names=target_names,
                                filled=True, rounded=True,
                                special_characters=True)  
# Capturo el source
graph = graphviz.Source(dot_data)
# Muestro el contenido de graph
graph

In [13]:
# Vamos a testear el modelo
print("Accuracy:",metrics.accuracy_score(y_test, y_pred))

Accuracy: 0.9523809523809523


Podemos observar que obtenemos un arbol con una precision del 90% de los animales clasificados de forma correcta.

**- ¿Varía ese resultado si se elimina el atributo “animal”? ¿Por
qué?**

In [14]:
feature_names2 = list(data3.columns)
feature_names2.remove('type')
#eliminamos el atributo animal
feature_names2.remove('animal')
x3 = data3[feature_names2]
y3 = data3.type
le.fit(data3['type'])
target_names2=le.classes_

X_train, X_test, y_train, y_test = train_test_split(x3, y3, random_state=0, test_size=0.2)
arbol3 = tree.DecisionTreeClassifier(criterion = "entropy")
arbol3 = arbol3.fit(X_train,y_train)
y_pred = arbol3.predict(X_test)
dot_data = tree.export_graphviz(arbol3, out_file=None,feature_names=feature_names2,class_names=target_names2,filled=True, rounded=True,special_characters=True)  
#graph = graphviz.Source(dot_data)
#graph
print("Accuracy:",metrics.accuracy_score(y_test, y_pred))

Accuracy: 0.9523809523809523


Al imprimir el accuracy vemos que varia mucho al eliminar el atributo indicado, ya que pasa a tener una prediccion de un 5% mas que el anterior.


**- Cuantos niveles posee el árbol generado? ¿Qué atributos
debemos modificar si deseamos realizar una poda del mismo?
Modifique esos atributos para que el árbol generado conste de
4 niveles. ¿Afecta la eficiencia de la clasificación esta
modificación?**


In [0]:

# Features
feature_names = list(data3.columns)
feature_names.remove('type')
x3 = data3[feature_names]
# Target
y3 = data3.type
le.fit(data3['type'])
target_names=le.classes_
#armo el arbol separando en 80-20 entrenamiento y testeo
X_train, X_test, y_train, y_test = train_test_split(x3, y3, random_state=0, test_size=0.2)
arbol3 = tree.DecisionTreeClassifier(criterion = "entropy", max_depth=3)
arbol3 = arbol3.fit(X_train,y_train)
y_pred = arbol3.predict(X_test)
dot_data = tree.export_graphviz(arbol3, out_file=None,feature_names=feature_names,class_names=target_names,filled=True, rounded=True,special_characters=True)  
graph = graphviz.Source(dot_data)
graph

In [16]:
print("Accuracy:",metrics.accuracy_score(y_test, y_pred))

Accuracy: 0.8571428571428571


-Podemos observar en el primer caso que obtenemos un arbol con 7 niveles. 

-Para modificar y que conste de 4 niveles basta con restringirselo en los parametros pasados al armar el arbol, mas precisamente el "max_depth" al cual setearemos con el valor indicado.

-Esta decision afecta la clasificacion ya que estamos forzando al arbol a tener una extension acotada a la que naturalmente tendria, por ende bajara su precision, especificamente para este caso en casi un 5% con respecto al arbol original. 




---


## PUNTO 4

**Consiga: Se provee la base de datos de los pasajeros del famoso barco que se hundiera en su viaje inaugural (archivo titanic-en.csv) con los siguientes atributos y valores posibles.**

**Genere el árbol de clasificación, explore la solución dada y las posibles alternativas para obtener un nuevo árbol que clasifique “mejor”.**

In [0]:
!rm titanic-en.csv
!wget https://raw.githubusercontent.com/jumafernandez/BDM/master/TPs/TP05/TP0501/titanic-en.csv

# Lee el archivo
data4 = pd.read_csv("titanic-en.csv") 
# Preview the first 5 lines of the loaded data 
data4.head()

In [0]:

le = preprocessing.LabelEncoder()
for column_name in data4.columns:
  if (data4[column_name].dtype == object) & (column_name!='survived'):
    data4[column_name] = le.fit_transform(data4[column_name])
    
# Features
feature_names = list(data4.columns)
# Elimino survived porque es la clase
feature_names.remove('survived')
x4 = data4[feature_names]
# Target
y4 = data4.survived
le.fit(data4['survived'])
target_names=le.classes_
#armo el arbol separando en 80-20 entrenamiento y testeo
X_train, X_test, y_train, y_test = train_test_split(x4, y4, random_state=0, test_size=0.2)
# Creo el objeto Decision Tree classifer
arbol4 = tree.DecisionTreeClassifier()
# Entreno el Decision Tree Classifer
arbol4 = arbol4.fit(X_train,y_train)
#Realizo las predicciones en función del árbol generado
y_pred = arbol4.predict(X_test)
# graficamos el arbol
dot_data = tree.export_graphviz(arbol4, out_file=None,
                                feature_names=feature_names,
                                class_names=target_names,
                                filled=True, rounded=True,
                                special_characters=True)  
# Capturo el source
graph = graphviz.Source(dot_data)
# Muestro el contenido de graph
graph

In [19]:
print("Accuracy:",metrics.accuracy_score(y_test, y_pred))

Accuracy: 0.7732426303854876


Para obtener una mejor clasificacion podemos modifcarlo de la siguiente manera:

In [0]:
feature_names = list(data4.columns)
feature_names.remove('survived')
x4 = data4[feature_names]
y4 = data4.survived
le.fit(data4['survived'])
target_names=le.classes_

X_train, X_test, y_train, y_test = train_test_split(x4, y4, random_state=0, test_size=0.4)
# Moifico parametros.
arbol4 = tree.DecisionTreeClassifier(criterion = "gini",min_samples_leaf=9)
arbol4 = arbol4.fit(X_train,y_train)
y_pred = arbol4.predict(X_test)

dot_data = tree.export_graphviz(arbol4, out_file=None,feature_names=feature_names,class_names=target_names,filled=True, rounded=True,special_characters=True)  
graph = graphviz.Source(dot_data)
graph

In [21]:
print("Accuracy:",metrics.accuracy_score(y_test, y_pred))

Accuracy: 0.7911464245175936


Duplicando el area de Test, modificando el criterio y aumentando el numero minimo de tuplas para formar una hoja se logro aumentar la presicion en un 2% con un total de 79,11%.



---


## PUNTO 5

**Consigna: Un Banco de Portugal realizó una campaña de marketing en busca de clientes de plazos fijos basada en llamados telefónicos. Se provee el dataset real (bank-full.csv) con más 45000 instancias y el detalle (bank-names.txt) de los datos registrados de cada una de las personas contactadas por la entidad bancaria.**

**5 a. Realice las tareas necesarias para poder procesar el dataset en
Scikit-Learn.**

In [0]:
!rm bank-full.csv
!wget https://raw.githubusercontent.com/jumafernandez/BDM/master/TPs/TP05/TP0501/bank-full.csv

# Lee el archivo
data5 = pd.read_csv("bank-full.csv", sep=";") 
# Preview the first 5 lines of the loaded data 
data5.head()

**5 b. Luego, genere el árbol de decisión, y optimice los resultados, con el objetivo de explicar cuáles son las características más importantes que permiten identificar a una persona que accederá o no al plazo fijo. Documente los resultados.**

In [0]:
le = preprocessing.LabelEncoder()
for column_name in data5.columns:
  if (data5[column_name].dtype == object) & (column_name!='y'):
    data5[column_name] = le.fit_transform(data5[column_name])

# Features
feature_names = list(data5.columns)
feature_names.remove('y')
x5 = data5[feature_names]
# Target
y5 = data5.y
le.fit(data5['y'])
target_names=le.classes_

# Separo en 70-30 entrenamiento y testeo
X_train, X_test, y_train, y_test = train_test_split(x5, y5, random_state=0, test_size=0.2)
# Creo el objeto Decision Tree classifer
arbol5 = tree.DecisionTreeClassifier(criterion='gini', max_depth=15)
# Entreno el Decision Tree Classifer
arbol5 = arbol5.fit(X_train,y_train)
#Realizo las predicciones en función del árbol generado
y_pred = arbol5.predict(X_test)
# graficamos el arbol
dot_data = tree.export_graphviz(arbol5, out_file=None,
                                feature_names=feature_names,
                                class_names=data5.y,
                                filled=True, rounded=True,
                                special_characters=True)  
# Capturo el source
graph = graphviz.Source(dot_data)
# Muestro el contenido de graph
graph

In [0]:
# Vamos a testear el modelo
print("Accuracy:",metrics.accuracy_score(y_test, y_pred))
# Vemos un reporte de clasificación de varias métricas
print(metrics.classification_report(y_test, y_pred))
print("cantidad de niveles: " + str(arbol5.tree_.max_depth))

Para optimizar los resultados se optó luego de una primera prueba, por limitar el numero de niveles maximos del arbol a 15 (inicialmente era de 31), cambiar el criterio a "Gini".

Tambien se concluyo que una de las caracteristicas mas importantes para determinar el resultado es la duration de la llamada, por lo cual si ponemos esta linea cuando cargamos los features:

In [0]:
feature_names.remove('duration')

El acurracy decaera casi un 7% con respecto al 88% inicial.



---


## PUNTO 6

**Consiga: Se requiere clasificar a los aspirantes 2014 que abandonaron la Universidad al primer cuatrimestre del 2015 en función de las características socio-económicas recolectadas en una encuesta realizada al momento de inscripción y volcada en el dataset encuesta_universitaria.csv que aún no está depurado.**

In [0]:
!wget https://raw.githubusercontent.com/jumafernandez/BDM/master/TPs/TP05/TP0501/encuesta_universitaria.csv
data6 = pd.read_csv("encuesta_universitaria.csv", encoding='latin_1', quotechar="'") 

## Reemplazo Nan por Cero
data6.loc[np.isnan(data6.Tiempo_Traslado), ['Tiempo_Traslado']] = 0
data6.loc[np.isnan(data6.Cantidad_Grupo_Familiar), ['Cantidad_Grupo_Familiar']] = 0
data6.loc[np.isnan(data6.Cantidad_Grupo_Familiar), ['Conocimientos_Procesador_Texto']] = 0
data6.loc[np.isnan(data6.Cantidad_Grupo_Familiar), ['Conocimientos_Planilla_Calculo']] = 0
data6.loc[np.isnan(data6.Cantidad_Grupo_Familiar), ['Conocimientos_Internet']] = 0

data6.head()

**6 a. Divida el dataset en dos partes, con proporciones del 80% y 20% del
total de las instancias del dataset.**

In [0]:
	
#target_names = ['ESTADO 1-2015']
feature_names = list(data6.columns)
feature_names.remove('ESTADO 1-2015')
# Features : convierte las variables categoricas en columnas con binarios como valores posibles.
x6 = pd.get_dummies(data6[feature_names])
# Target
y6 = data6['ESTADO 1-2015']
le.fit(data6['ESTADO 1-2015'])
target_names=le.classes_
# Separo en 70-30 entrenamiento y testeo
X_train, X_test_original, y_train, y_test_original = train_test_split(x6, y6, random_state=0, test_size=0.2)
# Creo el objeto Decision Tree classifer
arbol6 = tree.DecisionTreeClassifier(criterion='entropy')
# Entreno el Decision Tree Classifer
arbol6 = arbol6.fit(X_train,y_train)
#Realizo las predicciones en función del árbol generado
y_pred = arbol6.predict(X_test_original)
# graficamos el arbol
dot_data = tree.export_graphviz(arbol6, out_file=None,
                                feature_names=x6.columns,
                                class_names=target_names,
                                filled=True, rounded=True,
                                special_characters=True)  
# Capturo el source
graph = graphviz.Source(dot_data)
# Muestro el contenido de graph
graph

In [44]:
print("Accuracy:",metrics.accuracy_score(y_test_original, y_pred))

Accuracy: 0.6106870229007634


**6 b. Utilice el primer conjunto de instancias para generar 10 árboles
diferentes con distintas configuraciones y conserve los resultados.**

**Arbol Alternativo 1**

In [28]:
X_train, X_test, y_train, y_test = train_test_split(x6, y6, random_state=0, test_size=0.2)
arbol6 = tree.DecisionTreeClassifier(criterion='entropy',max_depth=25)
arbol6 = arbol6.fit(X_train,y_train)
y_pred1 = arbol6.predict(X_test_original)
dot_data = tree.export_graphviz(arbol6, out_file=None,feature_names=x6.columns,class_names=target_names,filled=True, rounded=True,special_characters=True)  

print("Accuracy:",metrics.accuracy_score(y_test_original, y_pred1))

Accuracy: 0.6225614927905004


**Arbol Alternativo 2**

In [29]:
X_train, X_test, y_train, y_test = train_test_split(x6, y6, random_state=0, test_size=0.2)
arbol6 = tree.DecisionTreeClassifier(criterion='entropy',max_depth=5)
arbol6 = arbol6.fit(X_train,y_train)
y_pred2 = arbol6.predict(X_test_original)
dot_data = tree.export_graphviz(arbol6, out_file=None,feature_names=x6.columns,class_names=target_names,filled=True, rounded=True,special_characters=True)  

print("Accuracy:",metrics.accuracy_score(y_test_original, y_pred2))

Accuracy: 0.6887192536047498


**Arbol Alternativo 3**

In [30]:
X_train, X_test, y_train, y_test = train_test_split(x6, y6, random_state=0, test_size=0.4)
arbol6 = tree.DecisionTreeClassifier(criterion='gini',min_samples_leaf=15)
arbol6 = arbol6.fit(X_train,y_train)
y_pred3 = arbol6.predict(X_test_original)
dot_data = tree.export_graphviz(arbol6, out_file=None,feature_names=x6.columns,class_names=target_names,filled=True, rounded=True,special_characters=True)  

print("Accuracy:",metrics.accuracy_score(y_test_original, y_pred3))

Accuracy: 0.6522476675148431


**Arbol Alternativo 4**

In [31]:
X_train, X_test, y_train, y_test = train_test_split(x6, y6, random_state=0, test_size=0.1)
arbol6 = tree.DecisionTreeClassifier(criterion='gini',min_samples_leaf=40)
arbol6 = arbol6.fit(X_train,y_train)
y_pred4 = arbol6.predict(X_test_original)
dot_data = tree.export_graphviz(arbol6, out_file=None,feature_names=x6.columns,class_names=target_names,filled=True, rounded=True,special_characters=True)  

print("Accuracy:",metrics.accuracy_score(y_test_original, y_pred4))

Accuracy: 0.6895674300254453


**Arbol Alternativo 5**

In [32]:
X_train, X_test, y_train, y_test = train_test_split(x6, y6, random_state=0, test_size=0.2)
arbol6 = tree.DecisionTreeClassifier(criterion='gini',min_samples_leaf=10,max_leaf_nodes=30)
arbol6 = arbol6.fit(X_train,y_train)
y_pred5 = arbol6.predict(X_test)
dot_data = tree.export_graphviz(arbol6, out_file=None,feature_names=x6.columns,class_names=target_names,filled=True, rounded=True,special_characters=True)  

print("Accuracy:",metrics.accuracy_score(y_test, y_pred5))

Accuracy: 0.6717557251908397


**Arbol Alternativo 6**

In [33]:
X_train, X_test, y_train, y_test = train_test_split(x6, y6, random_state=0, test_size=0.3)
arbol6 = tree.DecisionTreeClassifier(criterion='entropy',max_depth=10,min_samples_leaf=15,max_leaf_nodes=20)
arbol6 = arbol6.fit(X_train,y_train)
y_pred6 = arbol6.predict(X_test_original)
dot_data = tree.export_graphviz(arbol6, out_file=None,feature_names=x6.columns,class_names=target_names,filled=True, rounded=True,special_characters=True)  

print("Accuracy:",metrics.accuracy_score(y_test_original, y_pred6))

Accuracy: 0.6878710771840543


**Arbol Alternativo 7**

In [34]:
X_train, X_test, y_train, y_test = train_test_split(x6, y6, random_state=0, test_size=0.2)
arbol6 = tree.DecisionTreeClassifier(criterion='entropy',max_depth=20,splitter='random',min_samples_leaf=100)
arbol6 = arbol6.fit(X_train,y_train)
y_pred7 = arbol6.predict(X_test_original)
dot_data = tree.export_graphviz(arbol6, out_file=None,feature_names=x6.columns,class_names=target_names,filled=True, rounded=True,special_characters=True)  

print("Accuracy:",metrics.accuracy_score(y_test_original, y_pred7))

Accuracy: 0.6802374893977947


**Arbol Alternativo 8**

In [35]:
X_train, X_test, y_train, y_test = train_test_split(x6, y6, random_state=0, test_size=0.4)
arbol6 = tree.DecisionTreeClassifier(criterion='gini',splitter='random',min_samples_leaf=10,min_samples_split=5)
arbol6 = arbol6.fit(X_train,y_train)
y_pred8 = arbol6.predict(X_test_original)
dot_data = tree.export_graphviz(arbol6, out_file=None,feature_names=x6.columns,class_names=target_names,filled=True, rounded=True,special_characters=True)  

print("Accuracy:",metrics.accuracy_score(y_test_original, y_pred8))

Accuracy: 0.628498727735369


**Arbol Alternativo 9**

In [36]:
X_train, X_test, y_train, y_test = train_test_split(x6, y6, random_state=0, test_size=0.5)
arbol6 = tree.DecisionTreeClassifier(criterion='gini',max_features=7)
arbol6 = arbol6.fit(X_train,y_train)
y_pred9 = arbol6.predict(X_test_original)
dot_data = tree.export_graphviz(arbol6, out_file=None,feature_names=x6.columns,class_names=target_names,filled=True, rounded=True,special_characters=True)  

print("Accuracy:",metrics.accuracy_score(y_test_original, y_pred9))

Accuracy: 0.6013570822731128


**Arbol Alternativo 10**

In [37]:
X_train, X_test, y_train, y_test = train_test_split(x6, y6, random_state=0, test_size=0.2)
arbol6 = tree.DecisionTreeClassifier(criterion='entropy',random_state=49,class_weight='balanced')
arbol6 = arbol6.fit(X_train,y_train)
y_pred10 = arbol6.predict(X_test_original)
dot_data = tree.export_graphviz(arbol6, out_file=None,feature_names=x6.columns,class_names=target_names,filled=True, rounded=True,special_characters=True)  

print("Accuracy:",metrics.accuracy_score(y_test_original, y_pred10))

Accuracy: 0.6081424936386769


**6.c. Luego, genere un nuevo árbol aplicando, a partir de los resultados,
un sistema de votación en el que el valor de la variable objetivo del
árbol resultante estará dado por el valor más frecuente adoptado por
los árboles originales.**

In [0]:
#en y_pred_total estara la votacion resultante
y_pred_total=[]
#recorro todos los y_predN (que tienen la misma longitud)
for x in range(0,len(y_pred10)) :
  abandono = 0
  activo = 0
  #aumento los contadores a partir de lo que me da cada uno
  if (y_pred1[x]=='ACTIVO') :
    activo= activo + 1
  else:
    abandono = abandono +1
  
  if (y_pred2[x]=='ACTIVO') :
    activo= activo + 1
  else:
    abandono = abandono +1
    
  if (y_pred3[x]=='ACTIVO') :
    activo= activo + 1
  else:
    abandono = abandono +1  
  
  if (y_pred4[x]=='ACTIVO') :
    activo= activo + 1
  else:
    abandono = abandono +1
    
  if (y_pred5[x]=='ACTIVO') :
    activo= activo + 1
  else:
    abandono = abandono +1
  
  if (y_pred6[x]=='ACTIVO') :
    activo= activo + 1
  else:
    abandono = abandono +1
  
  if (y_pred7[x]=='ACTIVO') :
    activo= activo + 1
  else:
    abandono = abandono +1  
   
  if (y_pred8[x]=='ACTIVO') :
    activo= activo + 1
  else:
    abandono = abandono +1 
  
  if (y_pred9[x]=='ACTIVO') :
    activo= activo + 1
  else:
    abandono = abandono +1
  
  if (y_pred10[x]=='ACTIVO') :
    activo= activo + 1
  else:
    abandono = abandono +1
    #el mas votado sera el que ira para ese numero de indice en el y_pred_total.
  if (activo>=abandono) :
    y_pred_total.append('ACTIVO')
  else:
    y_pred_total.append('ABANDONO')
  
    

**6.d Valide el árbol resultante con el 20% restante del dataset original.**

In [39]:
#y_pred_total es la votacion de los 10 arboles alternativos.
print("Accuracy:",metrics.accuracy_score(y_test_original, y_pred_total))

Accuracy: 0.6810856658184903


**6.e Documente las acciones realizadas y el conocimiento más importante
generado a partir del árbol de decisión resultante.**

Luego de las acciones realizadas se concluye que en lugar de mejorar el porcentaje de Accuracy, se quedo en el rango esperado, el mismo de los 10 arboles alternativos (de 60 a 69) siendo este de 67. Podria este mejorar con configuraciones particulares del arbol que mejoren este promedio hecho por votacion.


---

## PUNTO 7
### Medidas de evaluación para técnicas de clasificación:

En función de la clasificación realizada, complete las siguientes actividades:

#### A. ACCURACY
**1. Ahora, calcule el accuracy de ambos modelos.**

Tratando por “ambos” a los dos principales árboles del punto 6, el accuracy de cada uno de ellos es:

Normal: 0.5911789652247668

Votación: 0.6810856658184903

**2. ¿Cómo se interpreta la métrica anterior?**

Podemos interpretar que a medida que se iteren más pruebas de árboles con distintas configuraciones y comparándolo con los datos reales, irá mejorando la eficiencia obteniendo un mejor árbol por cada vez.

**3. ¿Qué aporta el accuracy?**

Es el porcentaje de tuplas del conjunto de prueba (x_test_original) que fueron correctamente clasificadas por el modelo de árbol hecho en cada punto.


#### B. RECALL/PRECISION.
**1. Calcule las métricas recall y precisión para ambos modelos.**


In [56]:
# Calculo la presicion y recall para el y_pred original
print(precision_recall_fscore_support(y_test, y_pred, average='macro'))

# Calculo la presicion y recall para el y_pred_total, es decir, el resultado de la votacion
print(precision_recall_fscore_support(y_test, y_pred_total, average='macro'))

#los primeros dos resultados son la precision y el recall respextivamente.

(0.5711773128111679, 0.5683290641009622, 0.569168487801383, None)
(0.6481830688583695, 0.6170920279895039, 0.6199095953741366, None)



Entonces queda:

**-Árbol Original**

Precisión: 0.571 

Recall: 0.568

**-Árbol Votación**

Precisión: 0.648 

Recall: 0.617

**2. ¿Cuál es la diferencia entre ambas?**

Se observa una mejora en las dos medidas por parte de la votación. Para la precision de un 7% y para recall de un 5%.

**3. ¿Qué aspectos aborda cada una?**

La precisión es el porcentaje de todas las instancias relevantes dividido entre las instancias obtenidas.

El Recall es el porcentaje de resultados relevantes totales clasificados correctamente por el algoritmo. 


#### C. MATRIZ DE CONFUSIÓN.
**1. ¿En qué casos el modelo clasifica mal?**

Clasifica mal si los datos de entrada originales que se quieren clasificar tienen una desviación demasiado grande hacia la asignacion de una clase (de todas las posibles).
En ese caso la medida de evaluacion puede indicar un buen desempeño del arbol aunque no sea asi.

**Ejemplo:** 
*Si hay 90 muestras de una clase A y  10 de la clase B, la medida seguramente indicará que se hizo una clasificacion con una gran preferencia por la clase A. Si clasifica todas las muestras como clase A su precisión será del 90%, porcentaje que en principio puede resultar alto, pero que deja con un alto costo de error para con la otra clase, con la cual tuvo un 100% de error al no clasificar ninguna bien.*